### PR TITLE
Fix #9151 (cabal init logs at loglevel "Log")

### LINKED ISSUE #9151: cabal init logs at loglevel "Log"?
On 3.10, but appears to be introduced by @ptkato in 3.8: 

```
Add informative comments to each field in the cabal file. (y/n)? [default: y] n
[Log] Using cabal specification: 3.4
[Log] Creating fresh file LICENSE...
[Log] Creating fresh file CHANGELOG.md...
[Log] Creating fresh directory ./src...
[Log] Creating fresh file src/MyLib.hs...
[Log] Creating fresh directory ./app...
[Log] Creating fresh file app/Main.hs...
[Log] Creating fresh directory ./test...
[Log] Creating fresh file test/Main.hs...
[Log] Creating fresh file arithmetic.cabal...
[Warning] No synopsis given. You should edit the .cabal file and add one.
[Info] You may want to edit the .cabal file and add a Description field.
```

This is a quick change to `FileCreators.hs` and removing the ADT entry for "Log" in the loglevel type. This is a good first issue.

### PR DESCRIPTION
**Pull request by aleeusgr**
The actual ADT for severity should change because "Log" is useless as a datatype, arguably it shouldn't even be an entry in the ADT and the severity should be at `Info` for everything that's not a `Warn`.

* [ ] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.
* [x] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.

## QA notes:

when calling `cabal init` the user sees log entries categorized as Info, while entries categorized as Log should not appear.

### CODE REVIEW

File: changelog.d/pr-9346
Comment (andreabedini) [inline changelog.d/pr-9346:11]:
```suggestion
Remove "Log" as a log level in favour of "Info".
```

I think a simple entry like this is enough. No need to lecture users in a changelog :D (although I'd be happy to be lectured personally, I bookmarked the link).

File: cabal-install/src/Distribution/Client/Init/Types.hs
Comment (andreabedini) [inline cabal-install/src/Distribution/Client/Init/Types.hs:458]: Why do we have separate log levels (severities? are they terms interchangeable here?) for `Init`? 
If we don't want to use the ones in `Distribution.Verbosity` because that's in Cabal, we could have them somewhere in `Distribution.Client`

**Review by andreabedini — COMMENTED** [review]
I am in with this change. I left a couple of comments.

File: cabal-install/src/Distribution/Client/Init/Types.hs
Comment (aleeusgr) [reply cabal-install/src/Distribution/Client/Init/Types.hs:458]: Thank you for the food for thought.

File: changelog.d/pr-9346
Comment (aleeusgr) [reply changelog.d/pr-9346:11]: I apologize for the tone of how it came out: English isn't my native.

I'm most happy being lectured on how to make better contributions.

**Review by geekosaur — APPROVED** [review]
Not repeating the points that have already been raised.

File: cabal-install/src/Distribution/Client/Init/Types.hs
Comment (emilypi) [reply cabal-install/src/Distribution/Client/Init/Types.hs:458]: @andreabedini these logs are separate because it's good UX to have by default - it's not clear what is and is not actionable logging in the output from stdout without some form of loglevel. 

I'm amenable to following up with a PR that improves this by processing the verbosity to pick and choose the loglevel, but I wouldn't want them gone completely.

Comment (emilypi) [inline cabal-install/src/Distribution/Client/Init/Types.hs:458]: Let's not derive the `Show` instance here, and do the `displaySeverity` thing we discussed on the ticket.

Comment (aleeusgr) [reply cabal-install/src/Distribution/Client/Init/Types.hs:458]: Sure, let's do this

File: cabal-install/src/Distribution/Client/Init/Types.hs
Comment (aleeusgr) [inline cabal-install/src/Distribution/Client/Init/Types.hs:460]: I dunno where is a good place to put this function.

**Comment by aleeusgr** [comment]
Validation test is broken.

- [ ] what's a validation test?
- [ ] does it have to do with where I placed the function?
- [ ] is it something that will resolve itself?

File: cabal-install/src/Distribution/Client/Init/Types.hs
Comment (emilypi) [reply cabal-install/src/Distribution/Client/Init/Types.hs:460]: it's fine under it like that. Usually i prefer grouping functions with the relevant data types if it's a simple choice to make. More complex functionality is harder.

File: cabal-install/src/Distribution/Client/Init/Types.hs
Comment (emilypi) [inline cabal-install/src/Distribution/Client/Init/Types.hs:464]: this is a prime example of when to use a [Pattern Match](https://en.wikibooks.org/wiki/Haskell/Pattern_matching)! I would suggest the following: 

 ```suggestion
displaySeverity severity = case severity of 
   | Info -> "Info"
   | Warning -> "Warn"
   | Error -> "Err"
```

This is equivalent, but with more idiomatic code. It also doesn't involve `Eq`. Equivalently, you could also write: 

```haskell
displaySeverity :: Severity -> String
displaySeverity Info = "Info"
displaySeverity Warning = "Warn" 
displaySeverity Error = "Err"
```

Your choice.

**Comment by emilypi** [comment]
@aleeusgr Re: validation tests - i'll take a look in a bit to remind myself.

File: cabal-install/tests/IntegrationTests2/config/default-config
Comment (aleeusgr) [inline cabal-install/tests/IntegrationTests2/config/default-config:1]: this again!
@emilypi I think this is why validate fails: I pushed auto generated files with my commits.
I gotta look into git and find how to fix this.

Comment (aleeusgr) [inline cabal-install/tests/IntegrationTests2/config/default-config:1]: 🤦
Again. Gotta delete this somehow

**Comment by aleeusgr** [comment]
OK. Meanwhile I'll try to run the test locally.

File: cabal-install/src/Distribution/Client/Init/Types.hs
Comment (aleeusgr) [reply cabal-install/src/Distribution/Client/Init/Types.hs]: I chose case expression because I often navigate using search. Repeating words make this way less convenient.

Comment (ulysses4ever) [reply cabal-install/src/Distribution/Client/Init/Types.hs]: One more step I'd suggest is to consider using the `LambdaCase` GHC extension, which allows:
```
displaySeverity = \case 
  | Info -> "Info"
  | Warning -> "Warn"
  | Error -> "Err"
```

Comment (aleeusgr) [reply cabal-install/src/Distribution/Client/Init/Types.hs]: That's neat, I'll add the suggested change.

Thanks.

Comment (ulysses4ever) [reply cabal-install/src/Distribution/Client/Init/Types.hs]: Oh, my bad, sorry. I'm happy with either version tbh.

**Comment by aleeusgr** [comment]
All checks pass. Maybe guard expression caused validate test to fail. 

@emilypi please check QA note I added at the top of this PR.
I will add the change suggested by @ulysses4ever unless there are objections.

I need to 
- [ ] rebase the PR to the latest of master branch. 
- [ ] wait for second approval
- [ ] add `squash and merge` label.

**Comment by Mikolaj** [comment]
@aleeusgr: don't hesitate to mark the PR with help_needed and ping the Matrix channel if you feel blocked.

**Comment by aleeusgr** [comment]
Hi @Mikolaj 
I think we did everything that was planned here and checks pass.

I was waiting for the second approval, I don't know how things work here: I assumed someone would ping me if any actions are needed.

@emilypi could you confirm we achieved our goals for this PR?

Thanks.

**Comment by Mikolaj** [comment]
I was looking at your checklist in the comment above. Any ideas for an automated test?

**Comment by aleeusgr** [comment]
Maybe we could parse the log output line by line and fail if [Log] shows up?

**Comment by Mikolaj** [comment]
Sounds good to me. Perhaps also check the test output is not empty (and make sure the cabal-install commands issued in the test really should log something). Alternatively, there are golden tests somewhere (see the READMEs in the test folders and also CONTRIBUTING) so, if they apply to this situation, you can ensure the log output is always exactly the same as you record and accept initially. I suppose `cabal init` logs are deterministic, so that may work.

**Comment by aleeusgr** [comment]
- [ ] write/find a test that interacts with log output of `cabal-install`
- [ ] look at suites:
  - [ ] unit tests 
  - [ ] integration tests
  - [ ] long tests

**Comment by aleeusgr** [comment]
- [x] review golden tests
golden tests are stored in https://github.com/haskell/cabal/tree/master/cabal-install/tests/fixtures/init/golden
what is a [golden test](https://ro-che.info/articles/2017-12-04-golden-tests) ?
Golden test is like a unit test but its output is stored in a file.

**Comment by aleeusgr** [comment]
https://github.com/haskell/cabal/tree/master/cabal-install/tests#integration-tests

each test is a shell script.
🤔
But I don't see any .sh files, for example cabal-install/tests contains no shell scripts.

**Comment by aleeusgr** [comment]
- [ ] https://github.com/haskell/cabal/tree/master/cabal-testsuite#how-to-write
- [x] look at the tests that live in `cabal-testsuite/PackageTests`
  - [ ] build cabal-testsuite
  - [x] `nix develop github:yvan-sraka/cabal.nix`
- [x]  https://stackoverflow.com/questions/42702618/parsing-a-log-file-in-haskell
- [x] https://dev.to/anaynayak/haskell-parsing-a-log-message-1290
- [x] parse log?: https://hackage.haskell.org/package/attoparsec

**Comment by emilypi** [comment]
What you could do is probably add a parameter to add a log queue accumulator in the prompt state. I.e. 

```haskell
newtype PurePrompt a = PurePrompt
  { _runPrompt
      :: NonEmpty String
      -> Either BreakException (a, NonEmpty String)
  }
  deriving (Functor)
  ```
  
  Should turn into the following: 
  
  
  ```haskell
  
  type LogQueue = [Text]

  newtype PurePrompt a = PurePrompt
  { _runPrompt
      :: NonEmpty String
      -> Either BreakException (a, NonEmpty (LogQueue, String))
  }
  deriving (Functor)
```

Then the instances will need to be adjusted so that dedicated log functions are used which pipe to the correct behaviors in the IO case and the `PurePrompt` case. This is extracurricular. I think if you establish that the log output is doing the correct thing in a screenshot and raise an issue to track logging in `PurePrompt` state, I'll accept the PR.

**Comment by aleeusgr** [comment]
image

**Comment by aleeusgr** [comment]
https://github.com/haskell/cabal/issues/9482

File: cabal-install/src/Distribution/Client/Init/Types.hs
Comment (andreabedini) [reply cabal-install/src/Distribution/Client/Init/Types.hs]: Ok, that's reasonable, we had to introduce separate log levels that have a specific meaning to this phase :+1:

**Review by emilypi — APPROVED** [review]
Great job @aleeusgr :)

Congrats on contributing to Cabal!

**Comment by aleeusgr** [comment]
Thank you @emilypi
It's a pleasure 😀

**Comment by aleeusgr** [comment]
dear @geekosaur or @emilypi 

I regret to inform you that there is a problem with merging this PR:
image

I wonder what I could do to make it go through but it looks I am powerless in this situation.

Thank you for your attention

**Comment by Mikolaj** [comment]
Success! It must have been a mergify hiccup (and then a GHA random hangup, which I restarted).

@aleeusgr: congrats again! :)

**Pull request merged**
